Added LuGre friction model parameters to protocol

ale-git opened the pull request:
The yarprobotinterface parser is now able to read LuGre friction model parameters from robot configuration files.

pattacini commented:
Originated from https://github.com/robotology/icub-firmware-models/pull/97.

pattacini commented:
Reposting here thsi [comment](https://github.com/robotology/icub-firmware-models/pull/97#issuecomment-2431029151) as it pertains more to this PR:

>Hi @ale-git
>
>What's the strategy behind parsing the LuGre params?
>Ideally, to provide back-compatibility (quite important here), if the section is not found then the SW should NOT make use of the LuGre.

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.h:
```diff
@@ -297,6 +297,7 @@ class yarp::dev::embObjMotionControl:   public DeviceDriver,
     std::vector<eomc::timeouts_t>            _timeouts;
 
     std::vector<eomc::impedanceParameters_t> _impedance_params;   /** impedance parameters */ // TODO doubled!!! optimize using just one of the 2!!!
+    std::vector<eomc::lugreParameters_t>     _lugre_params;   /** impedance parameters */ // TODO doubled!!! optimize using just one of the 2!!!
```
Maybe this comment is due to a copy&paste option?

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.h:
```diff
@@ -510,6 +510,22 @@ typedef struct
     impedanceLimits_t limits;
 } impedanceParameters_t;
 
+typedef struct
```
Please add a comment for the new struct defined and for each params.

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.cpp:
```diff
@@ -2457,6 +2457,110 @@ bool Parser::parseImpedanceGroup(yarp::os::Searchable &config,std::vector<impeda
+    Bottle lugreGroup;
+    lugreGroup=config.findGroup("LUGRE","LUGRE parameters");
+
+    if(lugreGroup.isNull())
```
Are you sure you want to make the Lugre param mandatory?

pattacini replied:
Thanks @valegagge for spotting this.
Nope, we cannot make the presence of LuGre section mandatory.

My previous [comment](https://github.com/robotology/icub-main/pull/989#issuecomment-2435319879) goes in the same direction.

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.cpp:
```diff
@@ -1370,6 +1378,16 @@ bool embObjMotionControl::init()
         jconfig.impedance.offset    = 0;
+        
+        jconfig.lugre_params.Km     = _lugre_params[logico].Km;
+		jconfig.lugre_params.Kw     = _lugre_params[logico].Kw;
```
Please fixthe alignment of the code.

valegagge reviewed (changes requested):
Hi @ale-git, I wrote down some small changes. The most important one is related to making the new parameter group mandatory. This means that all robots' configurations need to be updated accordingly.

valegagge commented:
Hi @ale-git , 

in addition, I suggest updating the [template of the robot configuration](https://github.com/robotology/robots-configuration/tree/devel/iCubTemplates/iCubTemplateV6_0) files in order to maintain updated the documentation. If the new group is not mandatory you can add the info to the version 6.

Thanks

ale-git commented:
Hi @pattacini and @valegagge I've made the requested changes (LuGre section is not mandatory and the other issues).

pattacini commented:
Thanks @ale-git 

We still need to address https://github.com/robotology/icub-main/pull/989#pullrequestreview-2385144052.

For what concerns https://github.com/robotology/icub-main/pull/989#issuecomment-2460503246, I believe a subsequent PR will follow on `robots-configuration`, so that we can keep Templates up-to-date as well.

cc @valegagge

valegagge commented:
> Hi @pattacini and @valegagge I've made the requested changes (LuGre section is not mandatory and the other issues).

I'll do the review in the next few days (asap)

valegagge reviewed (approved):
Thanks @ale-git !

pattacini commented:
Well done.
We have to first merge https://github.com/robotology/icub-firmware-shared/pull/99#pullrequestreview-2385088941 and then put this in ready for review.

pattacini commented on conf/iCubFindDependencies.cmake:
```diff
@@ -64,7 +64,7 @@ checkandset_dependency(IPOPT)
 checkandset_dependency(Qt5)
 
-set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.39.1)
+set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.41.0)
```
Hi @ale-git 

This has to be updated to `1.40.1` as per https://github.com/robotology/icub-firmware-shared/pull/99?notification_referrer_id=NT_kwDOADkJ1rMxMjk4OTAxMTk4MzozNzM4MDcw#discussion_r1846898167.

cc @marcoaccame

pattacini replied:
Same as in https://github.com/robotology/icub-main/pull/989#discussion_r1846911095.

pattacini reviewed (approved):
See https://github.com/robotology/icub-main/pull/989/files#r1846904822.

marcoaccame commented on conf/iCubFindDependencies.cmake:
```diff
@@ -64,7 +64,7 @@ checkandset_dependency(IPOPT)
 checkandset_dependency(Qt5)
 
-set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.39.1)
+set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.41.0)
```
Hi @ale-git, pls set it to be 1.40.1 so that it matches what in here: https://github.com/robotology/icub-firmware-shared/pull/99#discussion_r1846898167

pattacini reviewed (changes requested):
See https://github.com/robotology/icub-main/pull/989#discussion_r1846911095.

pattacini reviewed (changes requested):
Hi @ale-git 

We need to address conflicts. In particular, we need to depend on the latest icub-firmware-shared version, which is `v1.41.0`.

pattacini commented:
/remind December 16, @ale-git please proceed asap with this.

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.cpp:
```diff
@@ -2457,6 +2457,115 @@ bool Parser::parseImpedanceGroup(yarp::os::Searchable &config,std::vector<impeda
+        {
+            lugre[j].S0 = 0.0;
+        }
+        
```
@ale-git: In case of missing `LUGRE` parameters, which is the correct initialization? When you check the return value of this function, you initialize all `km` fields to -1.0.

ale-git replied:
You're right, it must be Km = -1

ale-git replied:
You're right, it must be Km = -1

valegagge replied:
@ale-git , you set 1.0 and not -1.0. Are you sure?

ale-git replied:
Yes it must be -1. Maybe I need some holidays...

valegagge replied:
solved in the commit f1dc2ce4f10995e60f54f1e3a6cf2ebe36109704. Thanks @ale-git

valegagge commented on src/libraries/icubmod/embObjMotionControl/embObjMotionControl.cpp:
```diff
@@ -953,6 +955,16 @@ bool embObjMotionControl::fromConfig_Step2(yarp::os::Searchable &config)
+    ////// LUGRE PARAMETERS
+    if(! _mcparser->parseLugreGroup(config,_lugre_params))
+    {
+        yWarning() << "embObjMC " << getBoardInfo() << "LUGRE section: missing, sensorless torque estimation disabled";
```
here the function returns false only if a bad format of `LUGRE` group is found. The case of missing `LUGRE` group is already done in the function itself.

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.h:
```diff
@@ -510,6 +510,24 @@ typedef struct
+    double Fc_neg; // Coulomb friction constant (negative rotation)
+    double Fs_pos; // Stribeck friction constant (positive rotation)
+    double Fs_neg; // Stribeck friction constant (negative rotation)
+    double tbd0;   // for future use?
```
If you don't use this parameter, maybe it is better to remove them.

valegagge commented on src/libraries/icubmod/embObjMotionControl/eomcParser.h:
```diff
@@ -510,6 +510,24 @@ typedef struct
+
+// LuGre friction model parameters
+typedef struct
+{
```
@ale-git Please add the measuring unit in the comment of each field.

valegagge reviewed (changes requested):
Hi @ale-git , I left some notes on the code.
Let me know what do you think. Thanks

pattacini commented on conf/iCubFindDependencies.cmake:
```diff
@@ -64,7 +64,7 @@ checkandset_dependency(IPOPT)
 checkandset_dependency(Qt5)
 
-set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.39.1)
+set(MINIMUM_REQUIRED_icub_firmware_shared_VERSION 1.40.1)
```
The new version is 1.41.1 and it's already integrated so just remove this modification.

pattacini reviewed (changes requested):
Hi @ale-git 

Please, rebase your branch on top of the latest upstream devel and check the following:
- `icub-fw-shared` versioning: just remove the modification as requested inline.
- address @valegagge's comments

ale-git commented:
@pattacini @valegagge I've made the requested changes and rebased on devel.

pattacini reviewed (approved):
Thanks @ale-git 👍🏻 
Fine by me. Awaiting CI and @valegagge's final approval before merging.

pattacini reviewed (changes requested):
Hi @ale-git 

Please, address the CI failure.

pattacini commented:
Put it back to draft as https://github.com/robotology/icub-firmware-shared/pull/99 still has to be merged.

ale-git commented:
I've updated the icub_firmware_shared required version and fixed the -1 bug, ready for merge as soon as icub_firmware_shared is merged (ready for it).

Nicogene commented:
The icub-firmware PRs have been merged, once the CI is happy we can merge also this one
